ntp-proto/src/nts_record.rs

```diff
@@ -1323,119 +1335,118 @@ impl KeyExchangeServer {
         self.tls_connection.write_tls(wr)
     }
 
-    fn send_response(
-        &mut self,
-        protocol: ProtocolId,
-        algorithm: AeadAlgorithm,
-        keys: NtsKeys,
+    fn send_records(
+        tls_connection: &mut rustls::ServerConnection,
+        records: &[NtsRecord],
     ) -> std::io::Result<()> {
-        let records =
-            NtsRecord::server_key_exchange_records(protocol, algorithm, &self.keyset, keys);
-
         let mut buffer = Vec::with_capacity(1024);
-        for record in records.into_iter() {
+        for record in records.iter() {
             record.write(&mut buffer)?;
         }
 
-        self.tls_connection.writer().write_all(&buffer)?;
-        self.tls_connection.send_close_notify();
+        tls_connection.writer().write_all(&buffer)?;
+        tls_connection.send_close_notify();
 
         Ok(())
     }
 
-    pub fn progress(self) -> ControlFlow<Result<(), KeyExchangeError>, Self> {
-        match self.progress_help() {
-            ControlFlow::Continue(c) => ControlFlow::Continue(c),
-            ControlFlow::Break(b) => ControlFlow::Break(b.map(drop)),
+    fn send_error_record(tls_connection: &mut rustls::ServerConnection, error: &KeyExchangeError) {
+        use KeyExchangeError::*;
+
+        const UNRECOGNIZED_CRITICAL_RECORD: u16 = 0;
+        const BAD_REQUEST: u16 = 1;
+        const INTERNAL_SERVER_ERROR: u16 = 2;
+
+        let errorcode = match error {
+            UnrecognizedCriticalRecord => UNRECOGNIZED_CRITICAL_RECORD,
+            BadRequest => BAD_REQUEST,
+            InternalServerError | Io(_) => INTERNAL_SERVER_ERROR,
+            UnknownErrorCode(_)
+            | NoValidProtocol
+            | NoValidAlgorithm
+            | InvalidFixedKeyLength
+            | NoCookies
+            | Tls(_)
+            | Certificate(_)
+            | DnsName(_)
+            | IncompleteResponse => BAD_REQUEST,
+        };
+
+        if let Err(io) = Self::send_records(tls_connection, &[NtsRecord::Error { errorcode }]) {
+            tracing::debug!(key_exchange_error = ?error, io_error = ?io, "sending error record failed");
         }
     }
 
-    fn progress_help(mut self) -> ControlFlow<Result<Self, KeyExchangeError>, Self> {
+    pub fn progress(
+        mut self,
+    ) -> ControlFlow<Result<rustls::ServerConnection, KeyExchangeError>, Self> {
```

Comment:
this is a side-by-side sort of change I'm afraid. This is pretty nasty, further refactoring suggestions are welcome